libraries/chain_kv/include/b1/session/key_value.hpp

```diff
@@ -0,0 +1,157 @@
+// \param the_key A pointer to the key.
+// \param key_length The size of the memory pointed by the key.
+// \param the_value A pointer to the value.
+// \param value_length The size of the memory pointed by the value.
+// \param a The memory allocator used to managed the memory used by the key_value instance.
+// \returns A key_value instance.
+// \remarks This factory guarentees that the memory needed for the key and value will be contiguous in memory.
+template <typename Key, typename Value, typename Allocator>
+key_value make_kv(const Key* the_key, size_t key_length, const Value* the_value, size_t value_length, Allocator& a) {
+    auto total_key_length = key_length * sizeof(Key);
+    auto total_value_length = value_length * sizeof(Value);
+    return make_kv(reinterpret_cast<const void*>(the_key), total_key_length, reinterpret_cast<const void*>(the_value), total_value_length, a);
+}
+
+template <typename Key, typename Value>
+key_value make_kv(const Key* the_key, size_t key_length, const Value* the_value, size_t value_length) {
+    return make_kv(reinterpret_cast<const void*>(the_key), key_length * sizeof(Key), reinterpret_cast<const void*>(the_value), value_length * sizeof(Key));
+}
+
+inline key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length) {
+    auto kv = key_value{};
+    kv.m_key = make_shared_bytes(key, key_length);
+    kv.m_value = make_shared_bytes(value, value_length);
+    return kv;
+}
+
+
+// Instantiaties a key_value instance from the given key/value pointers.
+//
+// \tparam Allocator The memory allocator type used to manage the shared_bytes memory. This type mush implement the "memory allocator" concept.
+// \param key A pointer to the key.
+// \param key_length The size of the memory pointed by the key.
+// \param value A pointer to the value.
+// \param value_length The size of the memory pointed by the value.
+// \param a The memory allocator used to managed the memory used by the key_value instance.
+// \returns A key_value instance.
+// \remarks This factory guarentees that the memory needed for the key and value will be contiguous in memory.
+template <typename Allocator>
+key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length, Allocator& a) {
+    auto prefix_size = 2 * sizeof(uint64_t) + 2 * sizeof(size_t);
+    auto key_chunk_length = key_length == 0 ? key_length : key_length + prefix_size;
+    auto value_chunk_length = value_length == 0 ? value_length : value_length + prefix_size;
+    auto* chunk = reinterpret_cast<char*>(a->allocate(key_chunk_length + value_chunk_length));
+    
+    auto populate_bytes = [&a](const void* data, size_t length, char* chunk) mutable
+    {
+        auto result = shared_bytes{};
+        
+        if (!data || length == 0)
+        {
+          return result;
+        }
+        
+        result.m_chunk_start = reinterpret_cast<uint64_t*>(chunk);
```

Comment:
This chunk format needs to be documented and could very well be a reified structure at this point.  I don't see any reason why its packed size done this way would be different than that of a struct.  Meanwhile, if it is on a given platform we would be able to respect the alignment using standard keywords.

Reply:
The chunk returned from the allocate call is already aligned, at least from what I've seen.

Reply:
Injected a structured.